src/gaborkernels.jl

```diff
@@ -0,0 +1,323 @@
+
+    params = float.(promote(λ, θ, ψ, σ, γ))
+    T = typeof(params[1])
+
+    ax = _to_axes(size_or_axes)
+    all(map(length, ax) .> 0) || throw(ArgumentError("Kernel size should be positive: $size_or_axes"))
+    if any(r->5λ > length(r), ax)
+        expected_size = @. 5λ * length(ax)
+        @warn "for wavelength `λ=$λ` the expected kernel size is expected to be larger than $expected_size."
+    end
+
+    Gabor{Complex{T}, T, typeof(first(ax))}(ax, params...)
+end
+
+@inline Base.size(kern::Gabor) = map(length, kern.ax)
+@inline Base.axes(kern::Gabor) = kern.ax
+@inline function Base.getindex(kern::Gabor, x::Int, y::Int)
+    ψ, σ, γ = kern.ψ, kern.σ, kern.γ
+    s, c = kern.sc # sincos(θ)
+    λ_scaled = kern.λ_scaled # 2π/λ
+
+    xr = x*c + y*s
+    yr = -x*s + y*c
+    yr = γ * yr
+    return exp(-(xr^2 + yr^2)/(2σ^2)) * cis(xr*λ_scaled + ψ)
+end
+
+
+"""
+    LogGaborComplex(size_or_axes, ω, θ; σω=1, σθ=1, normalize=true)
+
+Generate the 2-D Log Gabor kernel in spatial space by `Complex(r, a)`, where `r` and `a`
+are the frequency and angular components, respectively.
+
+More detailed documentation and example can be found in the `r * a` version
+[`LogGabor`](@ref).
+"""
+struct LogGaborComplex{T, TP,R<:AbstractUnitRange} <: AbstractMatrix{T}
+    ax::Tuple{R,R}
+    ω::TP
+    θ::TP
+    σω::TP
+    σθ::TP
+    normalize::Bool
+
+    # cache values
+    freq_scale::Tuple{TP, TP} # only used when normalize is true
+    ω_denom::TP # 1/(2(log(σω/ω))^2)
+    θ_denom::TP # 1/(2σθ^2)
+    function LogGaborComplex{T,TP,R}(ax::Tuple{R,R}, ω::TP, θ::TP, σω::TP, σθ::TP, normalize::Bool) where {T,TP,R}
+        σω > 0 || throw(ArgumentError("`σω` should be positive: $σω"))
+        σθ > 0 || throw(ArgumentError("`σθ` should be positive: $σθ"))
+        ω_denom = 1/(2(log(σω/ω))^2)
+        θ_denom = 1/(2σθ^2)
+        freq_scale = map(r->1/length(r), ax)
+        new{T,TP,R}(ax, ω, θ, σω, σθ, normalize, freq_scale, ω_denom, θ_denom)
+    end
+end
+function LogGaborComplex(
+        size_or_axes::Tuple, ω::Real, θ::Real;
+        σω::Real=1, σθ::Real=1, normalize::Bool=true,
+    )
+    params = float.(promote(ω, θ, σω, σθ))
+    T = typeof(params[1])
+    ax = _to_axes(size_or_axes)
+    LogGaborComplex{Complex{T}, T, typeof(first(ax))}(ax, params..., normalize)
+end
+
+@inline Base.size(kern::LogGaborComplex) = map(length, kern.ax)
+@inline Base.axes(kern::LogGaborComplex) = kern.ax
+
+@inline function Base.getindex(kern::LogGaborComplex, x::Int, y::Int)
+    ω_denom, θ_denom = kern.ω_denom, kern.θ_denom
+    # Although in `getindex`, the computation is heavy enough that this runtime if-branch is
+    # harmless to the overall performance at all
+    if kern.normalize
+        # normalize: from reference [1] of LogGabor
+        # By changing division to multiplication gives about 5-10% performance boost
+        x, y = (x, y) .* kern.freq_scale
+    end
+    ω = sqrt(x^2 + y^2) # this is faster than hypot(x, y)
+    θ = atan(y, x)
```

Comment:
It is often the case that we can pre-compute the frequency `ω` and orientation `θ` and then call `kern[ω, θ]`. In this case, we might want to introduce a sphere index type to cache the computation and make the `getindex` dispatch on it.

```julia
for p in SphereIndices(img)
    # by doing this, we only need to compute the sphere index once
    kern1[p] * kern2[p]
end
```

Because some kernels are defined orientation-invariant, they wouldn't need the `θ` information at all. Thus each component of `SphereIndex` must be computed lazily so that we don't compute `θ`.

cc @Tokazama for a possible use case for https://github.com/JuliaArrays/SpatioTemporalTraits.jl/pull/11. This is kind of the opposite of the `CompositeSystem` you proposed.

Reply:
I've long wanted to be able to have a method for explicitly extracting spatial indices and this is a great practical example that motivates an implementation. We need some way of having a spatial system in the midst of color, time, etc. which users may end up creating a view of. The `CompositeSystem` proposal was an attempt to accommodate that, but if we did go in that direction we would still need to be able to extract a succinct representation of just the spatial system (e.g., `SphericalSystem`).

Reply:
We do already have `coords_spatial` and `ImageAxes.filter_space_axes`. I'm not a fan of color axes because having one destroys some of our most important abstractions (one array element = one pixel), but I recognize we may want to support such constructs slightly better than we do now.

Reply:
> We do already have `coords_spatial` and `ImageAxes.filter_space_axes`.

I thought those methods were only reliable if the type encoding the coordinate system is kept at the top level and it was a cartesian system, like we usually assume with AxisArray. Could we rely on that approach if the coordinate system was encoded in nested arrays or the spatial indices couldn't just be permuted or sliced (e.g., one dimension was for the radius and the other was for the azimuth)?

>  I'm not a fan of color axes because having one destroys some of our most important abstractions 

I agree, but it would be good to play nice with the machine learning community on that one.

Reply:
We can fix the nesting in the same way you've been doing it in ArrayInterface. Don't need a new trait that does the same thing, right?

Reply:
But yeah, we should provide better support for ML. Not by changing our algorithms, but by making it easier to represent their raw data in our framework. Making algorithms work for arrays that have a color channel is messy and error-prone, because there is no way to know how to interpret that channel unless it's annotated unambiguously. In which case, why not just change the representation to get rid of the color channel?

Reply:
So if we wanted to facilitate what was proposed we could do something like this

```julia
SpatialIndices(x) = SpatialIndices(CoordinateSystem(x), x)
SpatialIndices(cs::SphericalSystem, x) = SphereIndices{SphereIndex}(cs, indices_spatial(x))
```

We still use existing traits (`indices_spatial`), but we need more information to know which indices correspond to the radius, azimuth, and zenith so we can compose each `SphereIndex`.

Reply:
Because some of the computation can be non-trivial, can this design handle the task "I only want to iterate the first dimension of the system so if possible, don't do the extra computation for the second dimension"?

Reply:
If I'm understanding everything correctly here, you have two unique sets of coordinate systems with corresponding spatial indices/coordinates.
I would think that's sufficient to have a function like this:
```julia
function map_spatial_indices(dest, dest_cs::CartesianSystem, src, src_cs::SphericalSystem)
    ...
end
```
That function certianly isn't pretty and it would be nice if we could make it work through `getindex`.